src/lib/libmints/deriv.cc

```diff
@@ -747,14 +751,16 @@ SharedMatrix Deriv::compute()
     corr->add(x_contr_);
     corr->add(tpdm_contr_);
     if (reference_separate) {
-        gradient_->add(x_ref_contr_);
-        gradient_->add(opdm_ref_contr_);
-        gradient_->add(tpdm_ref_contr_);
-        SharedMatrix scf_gradient(gradient_->clone());
-        scf_gradient->set_name("Reference Gradient");
-        scf_gradient->print_atom_vector();
-        wfn_->reference_wavefunction()->set_gradient(scf_gradient);
-        corr->print_atom_vector();
+        if ( x_ref_contr_ && opdm_ref_contr_ && tpdm_ref_contr_) {
```

Comment:
Sorry, I only just saw this mention.  I can't remember the details exactly, but the hacky logic was introduced around the same time that the wavefunction class was starting to be used.  For DFMP2 gradients, the reference information is extracted from a wavefunction object.  For CC codes, the SCF part came from disk I think, because those objects were not proper subclasses of wavefunctions.  IIRC, the CI code folded the reference contributions into the OPDM and TPDM, so this hack was to allow all of the these calculation types to be handled with the same chunk of code, avoiding major rewrites.  Now that those major rewrites have been happening, I fully support re-working this legacy code to make it easier to understand; this should be possible now that all of our correlated methods have a somewhat common interface.  Nasty code segments like this are the growing pains that result from massaging many executables into one, unfortunately.